pkg/api/helper/imageextraction.go

```diff
@@ -0,0 +1,75 @@
+package helper
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/load"
+)
+
+// GetAllImageStreamTags returns a deduplicated and unsorted list of all ImageStreamTags
+// referenced anywhere in a config. It only returns their namespace and name and drops the
+// cluster field, as we plan to remove that.
+func GetAllImageStreamTags(config load.ByOrgRepo) ([]types.NamespacedName, error) {
+	result := sets.String{}
+	for _, org := range config {
+		for _, repo := range org {
+			for _, cfg := range repo {
+				imageStreamTagReferenceMapIntoSet(cfg.BaseImages, &result)
+				imageStreamTagReferenceMapIntoSet(cfg.BaseRPMImages, &result)
+				if cfg.BuildRootImage != nil && cfg.BuildRootImage.ImageStreamTagReference != nil {
+					result.Insert(imageStreamTagReferenceToString(*cfg.BuildRootImage.ImageStreamTagReference))
+				}
+
+				for _, rawStep := range cfg.RawSteps {
+					if rawStep.InputImageTagStepConfiguration != nil {
+						result.Insert(imageStreamTagReferenceToString(rawStep.InputImageTagStepConfiguration.BaseImage))
+					}
+					if rawStep.SourceStepConfiguration != nil {
+						result.Insert(imageStreamTagReferenceToString(rawStep.SourceStepConfiguration.ClonerefsImage))
+					}
+					// TODO: Is this correct?
+					if rawStep.OutputImageTagStepConfiguration != nil {
+						result.Insert(imageStreamTagReferenceToString(rawStep.OutputImageTagStepConfiguration.To))
+					}
+				}
+			}
+		}
+	}
+
+	var ret []types.NamespacedName
+	var errs []error
+	for _, item := range result.UnsortedList() {
+		split := strings.Split(item, "/")
+		if n := len(split); n != 2 {
+			errs = append(errs, fmt.Errorf("splitting string %q by '/' did not yield 2 but %d results", item, n))
+			continue
+		}
+		if split[0] == "" || split[1] == "" {
+			errs = append(errs, fmt.Errorf("namespace or name of item %q is empty", item))
+			continue
+		}
+		if res := strings.Split(split[1], ":"); len(res) != 2 || res[0] == "" || res[1] == "" {
+			errs = append(errs, fmt.Errorf("imagestreamtag name %q is not valid (must be <<name>>:<<tag>>)", split[1]))
+			continue
+		}
+		ret = append(ret, types.NamespacedName{Namespace: split[0], Name: split[1]})
+	}
```

Comment:
These values are created by functions in this module (unless they were originally empty in the configuration, but that should be validated outside this method), so panicking seems appropriate instead of propagating an error.

Moreover, we could map `imageStreamTagReferenceToString` values directly to `types.NamespacedName` (instead of `struct{}`), then simply collect the values here.

Reply:
> These values are created by functions in this module (unless they were originally empty in the configuration, but that should be validated outside this method), so panicking seems appropriate instead of propagating an error.

The case the error handling is supposed to handle is that we somehow get an invalid config, for example because our validation is buggy. While this should not happen in the first place I would prefer not to blow the whole controller up if it does.

The reason I didn't directly map them to `NamespacedName` is that I need them deduplicated

Reply:
Right, but I'm suggesting using a `map[string]types.NamespacedName` instead of the `map[string]struct{}` (i.e. `sets.String`) we are using. We'd still get unique values.